python/mxnet/model.py

```diff
@@ -297,10 +308,10 @@ def kv_updater(index, grad, weight):
                         # pull back the sum gradients, to the same locations.
                         kv.pull(index, grad_list, priority=-index)
                 if not update_on_kvstore:
-                    opt_list = opt_state_blocks[index]
-                    # optimizea
-                    for w, g, state in zip(arg_list, grad_list, opt_list):
-                        optimizer.update(index, w, g, state)
+                    for k, p in enumerate(zip(arg_list, grad_list)):
+                        w, g = p
+                        updater(index*num_device+k, g, w)
```

Comment:
Here is a better proposal. Let us call ```opt.get_updater``` for each device(create maybe a better name).

So we will have one updater closure for on device, and use each updater closure to update each device. Please do a bit more docs in opt.get_updater(create_updater) on how the closure should be used(i.e. create for one device a time)